### PR TITLE
Added new target support in Map , WeakMap, Set and WeakSet

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-container-object.c [hunk 1]
```diff
@@ -370,13 +370,20 @@ ecma_op_container_create (const ecma_value_t *arguments_list_p, /**< arguments l
...
-  ecma_object_t *object_p = ecma_create_object (ecma_builtin_get (proto_id),
-                                                sizeof (ecma_extended_object_t),
-                                                ECMA_OBJECT_TYPE_CLASS);
+  ecma_object_t *proto_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_new_target), proto_id);
 
+  if (JERRY_UNLIKELY (proto_p == NULL))
+  {
+    return ECMA_VALUE_ERROR;
+  }
+  ecma_object_t * object_p  = ecma_create_object (proto_p,
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-container-object.c:383]: No space after `*`;

File: jerry-core/ecma/operations/ecma-container-object.c [hunk 2]
```diff
@@ -370,13 +370,20 @@ ecma_op_container_create (const ecma_value_t *arguments_list_p, /**< arguments l
...
+                                                 sizeof (ecma_extended_object_t),
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-container-object.c:384]: Misaligned lines.

File: jerry-core/api/jerry.c [hunk 3]
```diff
@@ -4421,50 +4421,57 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
...
+  ecma_value_t constructor_value = ecma_op_object_get_by_magic_id (ecma_builtin_get (proto_id),
+                                                                   LIT_MAGIC_STRING_CONSTRUCTOR);
+  JERRY_CONTEXT (current_new_target) = ecma_get_object_from_value (constructor_value);
+  ecma_value_t container_value = ecma_op_container_create (arguments_list_p,
+                                                           arguments_list_len,
+                                                           lit_id,
+                                                           proto_id);
+  JERRY_CONTEXT (current_new_target) = NULL;
+  ecma_free_value (constructor_value);
+  return  container_value;
```
Comment (rerobika) [inline jerry-core/api/jerry.c:4474]: Double space.

File: jerry-core/api/jerry.c [hunk 4]
```diff
@@ -4421,50 +4421,57 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
...
+      proto_id = ECMA_BUILTIN_ID_WEAKSET_PROTOTYPE;
+      break;
     }
 #endif /* ENABLED (JERRY_ES2015_BUILTIN_WEAKSET) */
     default:
     {
       return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Invalid container type.")));
     }
   }
+  ecma_value_t constructor_value = ecma_op_object_get_by_magic_id (ecma_builtin_get (proto_id),
```
Comment (rerobika) [inline jerry-core/api/jerry.c:4465]: This operation may throw error and also the new.target is potentially can be set to an object when this API function called. So how about this?

```c
ecma_object_t * old_new_target_p = JERRY_CONTEXT (current_new_target);

if (old_new_target_p == NULL)
{
  JERRY_CONTEXT (current_new_target) = ecma_op_create_object_object_noarg();
}

// ecma_op_container_create ...

if (old_new_target_p == NULL)
{
  ecma_deref_object (JERRY_CONTEXT (current_new_target));
  JERRY_CONTEXT (current_new_target) = old_new_target;
}

JERRY_ASSERT (JERRY_CONTEXT (current_new_target) == old_new_target);
```

File: jerry-core/api/jerry.c [hunk 5]
```diff
@@ -4421,50 +4421,71 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
       return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p)));
     }
   }
-
+  lit_magic_string_id_t lit_id;
```
Comment (dbatyai) [inline jerry-core/api/jerry.c:4424]: Please add an empty line before this.

File: jerry-core/api/jerry.c [hunk 6]
```diff
@@ -4421,50 +4421,71 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
...
+  ecma_builtin_id_t proto_id;
+  ecma_builtin_id_t ctor_id;
```
Comment (dbatyai) [inline jerry-core/api/jerry.c:4426]: And an empty line after this.

File: jerry-core/ecma/operations/ecma-container-object.c [hunk 7]
```diff
@@ -370,13 +370,20 @@ ecma_op_container_create (const ecma_value_t *arguments_list_p, /**< arguments l
...
 
-  ecma_object_t *object_p = ecma_create_object (ecma_builtin_get (proto_id),
-                                                sizeof (ecma_extended_object_t),
-                                                ECMA_OBJECT_TYPE_CLASS);
+  ecma_object_t *proto_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_new_target), proto_id);
 
+  if (JERRY_UNLIKELY (proto_p == NULL))
+  {
+    return ECMA_VALUE_ERROR;
+  }
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-container-object.c:382]: Please add an empty line after the block.

File: jerry-core/api/jerry.c [hunk 8]
```diff
@@ -4422,49 +4422,72 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
...
+  }
+
+  ecma_value_t container_value = ecma_op_container_create (arguments_list_p,
+                                                           arguments_list_len,
+                                                           lit_id,
+                                                           proto_id);
+  if (old_new_target_p == NULL)
+  {
+    JERRY_CONTEXT (current_new_target) = NULL;
+    JERRY_CONTEXT (current_new_target) = old_new_target_p;
```
Comment (rerobika) [inline jerry-core/api/jerry.c:4486]: This assignment is enough.

File: jerry-core/ecma/operations/ecma-container-object.c (hunk 7 above)
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-container-object.c:382]: This request is still valid.

File: jerry-core/api/jerry.c [hunk 9]
```diff
@@ -4422,49 +4422,72 @@ jerry_create_container (jerry_container_type_t container_type, /**< Type of the
...
+  }
+
+  JERRY_ASSERT (JERRY_CONTEXT (current_new_target) == old_new_target_p);
```
Comment (rerobika) [inline jerry-core/api/jerry.c:4489]: Let me revise myself.

the  `if (old_new_target_p == NULL) {...}; JERRY_ASSERT(...)` can be substituted with `JERRY_CONTEXT (current_new_target) = old_new_target_p;`. Via this way, no branch operation is needed.

**Pull request merged**